File: cmd/agentjam/vault.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/spf13/cobra"
	"golang.org/x/term"

	"github.com/JamieDF/agentjam/internal/config"
	"github.com/JamieDF/agentjam/internal/errs"
	"github.com/JamieDF/agentjam/internal/vault"
	"github.com/JamieDF/agentjam/internal/vault/filestore"
)

func vaultCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "vault",
		Short: "Manage credentials",
		Long: `The credential vault stores secrets encrypted at rest. Agents reference
credentials by opaque URIs (vault://id) and never see the raw value. To use
a credential, the agent asks the vault to inject it via a template.

This command is the user-facing admin interface.`,
	}

	cmd.AddCommand(
		vaultInitCmd(),
		vaultUnlockCmd(),
		vaultLockCmd(),
		vaultListCmd(),
		vaultAddCmd(),
		vaultRemoveCmd(),
		vaultShowCmd(),
		vaultTemplateCmd(),
		vaultAuditCmd(),
	)

	return cmd
}

func openVault() (*filestore.FileStore, error) {
	return filestore.New(config.VaultPath())
}

// promptPassphrase asks the user for a passphrase without echoing.
func promptPassphrase(prompt string) (string, error) {
	fmt.Fprint(os.Stderr, prompt)
	fd := int(os.Stdin.Fd())
	b, err := term.ReadPassword(fd)
	fmt.Fprintln(os.Stderr)
	if err != nil {
		return "", errs.Wrap(err, "read passphrase")
	}
	return string(b), nil
}

func promptSecret(prompt string) (string, error) {
	fmt.Fprint(os.Stderr, prompt)
	fd := int(os.Stdin.Fd())
	b, err := term.ReadPassword(fd)
	fmt.Fprintln(os.Stderr)
	if err != nil {
		return "", errs.Wrap(err, "read secret")
	}
	return string(b), nil
}

func vaultInitCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "init",
		Short: "Initialize a new vault",
		RunE: func(cmd *cobra.Command, _ []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			state, err := v.LockState(cmd.Context())
			if err != nil {
				return err
			}
			if state != vault.Uninitialized {
				return errs.Wrap(errs.ErrAlreadyExists, "vault already exists; use `unlock` instead")
			}
			pass, err := promptPassphrase("Choose a passphrase: ")
			if err != nil {
				return err
			}
			confirm, err := promptPassphrase("Confirm passphrase: ")
			if err != nil {
				return err
			}
			if pass != confirm {
				return errs.Wrap(errs.ErrInvalid, "passphrases do not match")
			}
			if len(pass) < 8 {
				return errs.Wrap(errs.ErrInvalid, "passphrase must be at least 8 characters")
			}
			if err := v.Unlock(cmd.Context(), pass); err != nil {
				return err
			}
			fmt.Fprintln(cmd.OutOrStdout(), "Vault initialized.")
			return nil
		},
	}
}

func vaultUnlockCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "unlock",
		Short: "Unlock the vault",
		RunE: func(cmd *cobra.Command, _ []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			pass, err := promptPassphrase("Passphrase: ")
			if err != nil {
				return err
			}
			if err := v.Unlock(cmd.Context(), pass); err != nil {
				return err
			}
			fmt.Fprintln(cmd.OutOrStdout(), "Vault unlocked.")
			return nil
		},
	}
}

func vaultLockCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "lock",
		Short: "Lock the vault",
		RunE: func(cmd *cobra.Command, _ []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			if err := v.Lock(cmd.Context()); err != nil {
				return err
			}
			fmt.Fprintln(cmd.OutOrStdout(), "Vault locked.")
			return nil
		},
	}
}

func vaultListCmd() *cobra.Command {
	return &cobra.Command{
		Use:     "list",
		Short:   "List credentials",
		Aliases: []string{"ls"},
		RunE: func(cmd *cobra.Command, _ []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			creds, err := v.List(cmd.Context())
			if err != nil {
				return err
			}
			if len(creds) == 0 {
				fmt.Fprintln(cmd.OutOrStdout(), "No credentials. Add one with `agentjam vault add`.")
				return nil
			}
			fmt.Fprintf(cmd.OutOrStdout(), "%-30s %-10s %-10s %s\n", "ID", "USES", "TAGS", "DESCRIPTION")
			for _, c := range creds {
				tags := strings.Join(c.Tags, ",")
				fmt.Fprintf(cmd.OutOrStdout(), "%-30s %-10d %-10s %s\n",
					c.ID, c.UsesTotal, tags, c.Description)
			}
			return nil
		},
	}
}

func vaultAddCmd() *cobra.Command {
	var (
		desc string
		tags []string
	)
	cmd := &cobra.Command{
		Use:   "add <id>",
		Short: "Add a credential",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			if unlocked, _ := v.IsUnlocked(cmd.Context()); !unlocked {
				return errs.Wrap(errs.ErrUnauthorized, "vault is locked; run `agentjam vault unlock`")
			}
			value, err := promptSecret("Value: ")
			if err != nil {
				return err
			}
			c := vault.Credential{
				ID:          vault.ID(args[0]),
				Description: desc,
				Tags:        tags,
				Value:       value,
			}
			if err := v.Add(cmd.Context(), c); err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "Added credential %q.\n", c.ID)
			return nil
		},
	}
	cmd.Flags().StringVar(&desc, "description", "", "Credential description")
	cmd.Flags().StringSliceVar(&tags, "tag", nil, "Tag (repeatable)")
	return cmd
}

func vaultRemoveCmd() *cobra.Command {
	return &cobra.Command{
		Use:     "remove <id>",
		Short:   "Remove a credential",
		Aliases: []string{"rm"},
		Args:    cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			if err := v.Remove(cmd.Context(), vault.ID(args[0])); err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "Removed %q.\n", args[0])
			return nil
		},
	}
}

func vaultShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show <id>",
		Short: "Show credential metadata and templates",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			c, err := v.Get(cmd.Context(), vault.ID(args[0]))
			if err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "ID:          %s\n", c.ID)
			fmt.Fprintf(cmd.OutOrStdout(), "Description: %s\n", c.Description)
			fmt.Fprintf(cmd.OutOrStdout(), "Created:     %s\n", c.CreatedAt.Format("2006-01-02 15:04"))
			fmt.Fprintf(cmd.OutOrStdout(), "Last Used:   %s\n", c.LastUsed.Format("2006-01-02 15:04"))
			fmt.Fprintf(cmd.OutOrStdout(), "Uses Total:  %d\n", c.UsesTotal)
			if len(c.Tags) > 0 {
				fmt.Fprintf(cmd.OutOrStdout(), "Tags:        %s\n", strings.Join(c.Tags, ", "))
			}
			tmpls, err := v.ListTemplates(cmd.Context(), c.ID)
			if err != nil {
				return err
			}
			if len(tmpls) > 0 {
				fmt.Fprintln(cmd.OutOrStdout(), "\nTemplates:")
				for _, t := range tmpls {
					fmt.Fprintf(cmd.OutOrStdout(), "  - %s (%s)\n", t.Name, t.Injection.Mode())
				}
			}
			return nil
		},
	}
}

func vaultTemplateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "template",
		Short: "Manage injection templates",
	}
	cmd.AddCommand(
		vaultTemplateListCmd(),
		vaultTemplateAddCmd(),
		vaultTemplateRemoveCmd(),
	)
	return cmd
}

func vaultTemplateListCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list <cred-id>",
		Short: "List templates for a credential",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			tmpls, err := v.ListTemplates(cmd.Context(), vault.ID(args[0]))
			if err != nil {
				return err
			}
			for _, t := range tmpls {
				fmt.Fprintf(cmd.OutOrStdout(), "%s (%s)\n", t.Name, t.Injection.Mode())
			}
			return nil
		},
	}
}

func vaultTemplateAddCmd() *cobra.Command {
	var (
		name        string
		envJSON     string
		wrapperTmpl string
		filePath    string
		stdin       bool
	)
	cmd := &cobra.Command{
		Use:   "add <cred-id>",
		Short: "Add an injection template to a credential",
		Long: `Add an injection template. Exactly one injection mode must be specified:

  --env '{"KEY": "$value"}'                      env vars
  --wrapper 'curl -H "Auth: $value" {original}' command wrapper
  --file /tmp/secrets.txt                        write to file
  --stdin                                        pipe to stdin

Example:
  agentjam vault template add github/work \
    --name gh_cli_env \
    --env '{"GH_TOKEN": "$value"}'`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			inj := vault.Injection{}
			modes := 0
			if envJSON != "" {
				var env map[string]string
				if err := json.Unmarshal([]byte(envJSON), &env); err != nil {
					return errs.Wrap(err, "parse --env JSON")
				}
				inj.Env = env
				modes++
			}
			if wrapperTmpl != "" {
				inj.CommandWrapper = &vault.WrapperInjection{Template: wrapperTmpl}
				modes++
			}
			if filePath != "" {
				inj.File = &vault.FileInjection{Path: filePath}
				modes++
			}
			if stdin {
				inj.Stdin = &vault.StdinInjection{TrailingNewline: true}
				modes++
			}
			if modes != 1 {
				return errs.Wrap(errs.ErrInvalid, "exactly one injection mode required")
			}
			tmpl := vault.Template{Name: name, Injection: inj}
			if err := v.SetTemplate(cmd.Context(), vault.ID(args[0]), tmpl); err != nil {
				return err
			}
			fmt.Fprintf(cmd.OutOrStdout(), "Added template %q.\n", name)
			return nil
		},
	}
	cmd.Flags().StringVar(&name, "name", "", "Template name (required)")
	cmd.Flags().StringVar(&envJSON, "env", "", "Env injection as JSON")
	cmd.Flags().StringVar(&wrapperTmpl, "wrapper", "", "Command wrapper template")
	cmd.Flags().StringVar(&filePath, "file", "", "File injection path")
	cmd.Flags().BoolVar(&stdin, "stdin", false, "Stdin injection")
	_ = cmd.MarkFlagRequired("name")
	return cmd
}

func vaultTemplateRemoveCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "remove <cred-id> <template-name>",
		Short: "Remove a template",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			if err := v.RemoveTemplate(cmd.Context(), vault.ID(args[0]), args[1]); err != nil {
				return err
			}
			fmt.Fprintln(cmd.OutOrStdout(), "Removed.")
			return nil
		},
	}
}

func vaultAuditCmd() *cobra.Command {
	var (
		cred  string
		limit int
	)
	return &cobra.Command{
		Use:   "audit",
		Short: "Show audit log",
		RunE: func(cmd *cobra.Command, _ []string) error {
			v, err := openVault()
			if err != nil {
				return err
			}
			entries, err := v.AuditLog(cmd.Context(), vault.AuditOptions{
				Credential: vault.ID(cred),
				Limit:      limit,
			})
			if err != nil {
				return err
			}
			for _, e := range entries {
				status := "✓"
				if !e.Success {
					status = "✗"
				}
				fmt.Fprintf(cmd.OutOrStdout(), "%s %s [%s] %s -> %s\n",
					e.Timestamp.Format("2006-01-02 15:04:05"),
					status, e.Credential, e.Template, e.Caller)
			}
			return nil
		},
	}
}

File: src/agent_knots/cli/vault.py
"""`agent-knots vault` subcommands, including `vault template`."""

from __future__ import annotations

import typer

from agent_knots.cli._format import format_ts
from agent_knots.config import vault_dir
from agent_knots.vault.store import LockState, VaultStore

vault_app = typer.Typer(help="Manage the encrypted credential vault.", no_args_is_help=True)

# Store a global reference so subcommands can access the same store.
_vault_store: VaultStore | None = None


def _get_vault() -> VaultStore:
    global _vault_store
    if _vault_store is None:
        _vault_store = VaultStore(vault_dir())
    return _vault_store


@vault_app.command()
def init() -> None:
    """Initialize a new vault with a passphrase."""
    store = _get_vault()
    if store.lock_state == LockState.UNLOCKED:
        typer.echo("Vault is already initialized and unlocked.")
        return
    if store.lock_state == LockState.LOCKED:
        typer.echo("Vault is already initialized but locked. Use 'vault unlock'.")
        return

    passphrase = typer.prompt("Choose a passphrase", hide_input=True)
    confirm = typer.prompt("Confirm passphrase", hide_input=True)
    if passphrase != confirm:
        typer.echo("Passphrases do not match.")
        raise typer.Exit(1)

    store.unlock(passphrase)
    typer.echo("Vault initialized and unlocked.")


@vault_app.command()
def unlock() -> None:
    """Unlock the vault with a passphrase."""
    store = _get_vault()
    if store.lock_state == LockState.UNLOCKED:
        typer.echo("Vault is already unlocked.")
        return
    if store.lock_state == LockState.UNINITIALIZED:
        typer.echo("Vault is not initialized. Use 'vault init' first.")
        raise typer.Exit(1)

    passphrase = typer.prompt("Passphrase", hide_input=True)
    try:
        store.unlock(passphrase)
        typer.echo("Vault unlocked.")
    except ValueError as e:
        typer.echo(f"Error: {e}")
        raise typer.Exit(1)


@vault_app.command()
def lock() -> None:
    """Lock the vault."""
    store = _get_vault()
    if store.lock_state != LockState.UNLOCKED:
        typer.echo("Vault is not unlocked.")
        return
    store.lock()
    typer.echo("Vault locked.")


@vault_app.command()
def status() -> None:
    """Show vault status."""
    store = _get_vault()
    state = store.lock_state.value
    count = len(store.list_credentials())
    typer.echo(f"Status: {state}")
    typer.echo(f"Credentials: {count}")


@vault_app.command()
def add(
    cred_id: str = typer.Argument(..., help="Credential ID (e.g. 'github/work')."),
    value: str = typer.Option("", "--value", help="The secret value. If omitted, you'll be prompted."),
    description: str = typer.Option("", "--description", help="Human-readable note."),
    tag: list[str] = typer.Option([], "--tag", help="Tags for this credential (repeatable)."),
) -> None:
    """Add a new credential to the vault.

    Examples:
        agent-knots vault add github --value ghp_xxx --tag git
        agent-knots vault add openai --tag production   (prompts for value)
    """
    from agent_knots.vault.store import Credential

    store = _get_vault()
    if store.lock_state != LockState.UNLOCKED:
        typer.echo("Vault is locked. Use 'vault unlock' first.")
        raise typer.Exit(1)

    # If no value given on CLI, prompt securely.
    if not value:
        value = typer.prompt("Credential value", hide_input=True)
        confirm = typer.prompt("Confirm value", hide_input=True)
        if value != confirm:
            typer.echo("Values do not match.")
            raise typer.Exit(1)

    try:
        store.add_credential(Credential(
            id=cred_id,
            value=value,
            description=description,
            tags=list(tag),
        ))
        typer.echo(f"Credential '{cred_id}' added.")
    except ValueError as e:
        typer.echo(f"Error: {e}")
        raise typer.Exit(1)


@vault_app.command(name="list")
def list_credentials() -> None:
    """List all credentials in the vault (without values)."""
    store = _get_vault()
    creds = store.list_credentials()
    if not creds:
        typer.echo("No credentials stored.")
        return
    for c in creds:
        tags = ", ".join(c.tags) if c.tags else ""
        typer.echo(f"  {c.id:30s}  uses={c.uses_total:>4d}  {tags}")


@vault_app.command()
def show(cred_id: str = typer.Argument(..., help="Credential ID.")) -> None:
    """Show a credential's value (requires unlocked vault)."""
    store = _get_vault()
    if store.lock_state != LockState.UNLOCKED:
        typer.echo("Vault is locked. Use 'vault unlock' first.")
        raise typer.Exit(1)

    try:
        value = store.use_credential(cred_id, caller="user")
        typer.echo(value)
    except ValueError as e:
        typer.echo(f"Error: {e}")
        raise typer.Exit(1)


@vault_app.command()
def remove(cred_id: str = typer.Argument(..., help="Credential ID to remove.")) -> None:
    """Remove a credential from the vault."""
    store = _get_vault()
    if store.lock_state != LockState.UNLOCKED:
        typer.echo("Vault is locked. Use 'vault unlock' first.")
        raise typer.Exit(1)

    try:
        store.remove_credential(cred_id)
        typer.echo(f"Credential '{cred_id}' removed.")
    except ValueError as e:
        typer.echo(f"Error: {e}")
        raise typer.Exit(1)


@vault_app.command()
def audit(
    cred_id: str = typer.Option("", "--credential", help="Filter by credential ID."),
    limit: int = typer.Option(0, "--limit", help="Max entries to show."),
) -> None:
    """Show the audit log."""
    from agent_knots.vault.store import AuditOptions

    store = _get_vault()
    entries = store.audit_log(AuditOptions(credential=cred_id, limit=limit))
    if not entries:
        typer.echo("No audit entries.")
        return
    for e in entries:
        status = "✓" if e.success else "✗"
        ts = format_ts(e.timestamp)
        typer.echo(f"  {ts}  {status}  {e.credential:25s}  {e.caller:15s}  {e.template}")


# ── vault template commands ──────────────────────────────────────────────────

vault_template_app = typer.Typer(
    help="Manage credential injection templates.", no_args_is_help=True
)
vault_app.add_typer(vault_template_app, name="template")


@vault_template_app.command(name="add")
def template_add(
    cred_id: str = typer.Argument(..., help="Credential ID (e.g. 'github/work')."),
    name: str = typer.Option(..., "--name", help="Template name (e.g. 'gh_cli_env')."),
    description: str = typer.Option("", "--description", help="Human-readable note."),
    env: str = typer.Option(
        "", "--env", help='JSON object of env vars to inject, e.g. \'{"GH_TOKEN": "$value"}\'.'
    ),
    file_path: str = typer.Option(
        "", "--file", help="Write the value to this path instead of an env var."
    ),
    file_permissions: str = typer.Option(
        "600", "--file-permissions", help="Octal permissions for --file (default 600)."
    ),
    stdin: bool = typer.Option(False, "--stdin", help="Pipe the value to the command's stdin."),
    wrapper: str = typer.Option(
        "", "--wrapper", help="Command wrapper template using {original} and $value."
    ),
) -> None:
    """Add or replace an injection template on a credential.

    Examples:
        agent-knots vault template add github/work --name gh_cli_env --env '{"GH_TOKEN": "$value"}'
        agent-knots vault template add github/work --name curl_bearer --wrapper "curl -H 'Authorization: token $value' {original}"
    """
    import json

    from agent_knots.vault.store import InjectionTemplate

    store = _get_vault()
    if store.lock_state != LockState.UNLOCKED:
        typer.echo("Vault is locked. Use 'vault unlock' first.")
        raise typer.Exit(1)

    try:
        env_dict = json.loads(env) if env else {}
        if not isinstance(env_dict, dict):
            raise ValueError("--env must be a JSON object")
    except (json.JSONDecodeError, ValueError) as e:
        typer.echo(f"Error: invalid --env: {e}")
        raise typer.Exit(1)

    try:
        store.set_template(cred_id, InjectionTemplate(
            name=name,
            description=description,
            env=env_dict,
            file_path=file_path or None,
            file_permissions=int(file_permissions, 8),
            stdin=stdin,
            command_wrapper=wrapper or None,
        ))
        typer.echo(f"Template '{name}' added to '{cred_id}'.")
    except ValueError as e:
        typer.echo(f"Error: {e}")
        raise typer.Exit(1)


@vault_template_app.command(name="list")
def template_list(cred_id: str = typer.Argument(..., help="Credential ID.")) -> None:
    """List templates on a credential."""
    store = _get_vault()
    templates = store.list_templates(cred_id)
    if not templates:
        typer.echo("No templates.")
        return
    for t in templates:
        kind = "env" if t.env else "file" if t.file_path else "stdin" if t.stdin else "wrapper"
        typer.echo(f"  {t.name:20s}  {kind:8s}  {t.description}")


@vault_template_app.command(name="show")
def template_show(
    cred_id: str = typer.Argument(..., help="Credential ID."),
    name: str = typer.Argument(..., help="Template name."),
) -> None:
    """Show a template's injection config."""
    store = _get_vault()
    t = store.get_template(cred_id, name)
    if t is None:
        typer.echo(f"Template {name!r} not found on {cred_id!r}.")
        raise typer.Exit(1)
    typer.echo(f"Name:        {t.name}")
    typer.echo(f"Description: {t.description or '—'}")
    if t.env:
        typer.echo(f"Env:         {t.env}")
    if t.file_path:
        typer.echo(f"File:        {t.file_path} (mode {oct(t.file_permissions)})")
    if t.stdin:
        typer.echo("Stdin:       yes")
    if t.command_wrapper:
        typer.echo(f"Wrapper:     {t.command_wrapper}")


@vault_template_app.command(name="remove")
def template_remove(
    cred_id: str = typer.Argument(..., help="Credential ID."),
    name: str = typer.Argument(..., help="Template name to remove."),
) -> None:
    """Remove a template from a credential."""
    store = _get_vault()
    try:
        store.remove_template(cred_id, name)
        typer.echo(f"Template '{name}' removed from '{cred_id}'.")
    except ValueError as e:
        typer.echo(f"Error: {e}")
        raise typer.Exit(1)
